exchanges/stream/websocket.go

```diff
@@ -0,0 +1,918 @@
+	defer w.connectionMutex.RUnlock()
+	return w.trafficMonitorRunning
+}
+
+func (w *Websocket) setConnectionMonitorRunning(b bool) {
+	w.connectionMutex.Lock()
+	w.connectionMonitorRunning = b
+	w.connectionMutex.Unlock()
+}
+
+// IsConnectionMonitorRunning returns status of connection monitor
+func (w *Websocket) IsConnectionMonitorRunning() bool {
+	w.connectionMutex.RLock()
+	defer w.connectionMutex.RUnlock()
+	return w.connectionMonitorRunning
+}
+
+func (w *Websocket) setDataMonitorRunning(b bool) {
+	w.connectionMutex.Lock()
+	w.dataMonitorRunning = b
+	w.connectionMutex.Unlock()
+}
+
+// IsDataMonitorRunning returns status of data monitor
+func (w *Websocket) IsDataMonitorRunning() bool {
+	w.connectionMutex.RLock()
+	defer w.connectionMutex.RUnlock()
+	return w.dataMonitorRunning
+}
+
+// CanUseAuthenticatedWebsocketForWrapper Handles a common check to
+// verify whether a wrapper can use an authenticated websocket endpoint
+func (w *Websocket) CanUseAuthenticatedWebsocketForWrapper() bool {
+	if w.IsConnected() && w.CanUseAuthenticatedEndpoints() {
+		return true
+	} else if w.IsConnected() && !w.CanUseAuthenticatedEndpoints() {
+		log.Infof(log.WebsocketMgr,
+			WebsocketNotAuthenticatedUsingRest,
+			w.exchangeName)
+	}
+	return false
+}
+
+// SetWebsocketURL sets websocket URL and can refresh underlying connections
+func (w *Websocket) SetWebsocketURL(url string, auth, reconnect bool) error {
+	defaultVals := url == "" || url == config.WebsocketURLNonDefaultMessage
+	if auth {
+		if defaultVals {
+			url = w.defaultURLAuth
+		}
+
+		err := checkWebsocketURL(url)
+		if err != nil {
+			return err
+		}
+		w.runningURLAuth = url
+
+		if w.verbose {
+			log.Debugf(log.WebsocketMgr,
+				"%s websocket: setting authenticated websocket URL: %s\n",
+				w.exchangeName,
+				url)
+		}
+
+		if w.AuthConn != nil {
+			w.AuthConn.SetURL(url)
+		}
+	} else {
+		if defaultVals {
+			url = w.defaultURL
+		}
+		err := checkWebsocketURL(url)
+		if err != nil {
+			return err
+		}
+		w.runningURL = url
+
+		if w.verbose {
+			log.Debugf(log.WebsocketMgr,
+				"%s websocket: setting unauthenticated websocket URL: %s\n",
+				w.exchangeName,
+				url)
+		}
+
+		if w.Conn != nil {
+			w.Conn.SetURL(url)
+		}
+	}
+
+	if w.IsConnected() && reconnect {
+		log.Debugf(log.WebsocketMgr,
+			"%s websocket: flushing websocket connection to %s\n",
+			w.exchangeName,
+			url)
+		return w.Shutdown()
+	}
+	return nil
+}
+
+// GetWebsocketURL returns the running websocket URL
+func (w *Websocket) GetWebsocketURL() string {
+	return w.runningURL
+}
+
+// Initialise verifies status and connects
+func (w *Websocket) Initialise() error {
+	if w.IsEnabled() {
+		if w.IsInit() {
+			return nil
+		}
+		return fmt.Errorf("%v websocket: already initialised", w.exchangeName)
+	}
+	w.setEnabled(w.enabled)
+	return nil
+}
+
+// SetProxyAddress sets websocket proxy address
+func (w *Websocket) SetProxyAddress(proxyAddr string) error {
+	if w.proxyAddr == proxyAddr {
```

Comment:
The needs to validate the url: `DEBUG] | EXCHANGE | 17/07/2020 14:12:12 | Kraken websocket: setting websocket proxy: lol`